### DISCUSSION TITLE
Eglot display more than first line of LSP doc string.  Fine control over Eglot's ElDoc functions.

### DISCUSSION THREAD

**Discussion by zino299792** [joaotavora/eglot#1467] [Q&A] [answered]
-   Server used: rust-analyzer
-   Emacs version: GNU Emacs 29.1 (build 1, x86<sub>64</sub>-apple-darwin22.1.0, Carbon Version 169 AppKit 2299) of 2024-02-05
-   Operating system: Mac OS 13.7
-   Eglot version: b014bca833a17f5b2258e88115f03cffa983d0bd
-   Eglot installation method: eglot-upgrade-eglot
-   Using Doom: No
-   Relevant Emacs settings
    
        (setq eldoc-documentation-strategy 'eldoc-documentation-compose)
        (setq eldoc-echo-area-use-multiline-p 'truncate-sym-name-if-fit)
        (setq eldoc-documentation-functions 'eglot-hover-eldoc-function)

After upgrading eglot from 1.12 to 1.17, the doc string in echo area is incomplete for `eglot-hover-eldoc-function` backend.

For example, when hoverring on a variant of `RustAnalyzerCmd`, only the complete path to the enum `rust_analyzer::cli::flags::RustAnalyzer` is displayed:

![2024-11-26_20-23-50_screenshot](https://github.com/user-attachments/assets/eadb50e2-e263-426e-b10d-bec5e61b46f3)

Definition of `RustAnalyzerCmd`:

```rust
#[derive(Debug)]
pub enum RustAnalyzerCmd {
    LspServer(LspServer),
    Parse(Parse),
    Symbols(Symbols),
    Highlight(Highlight),
    AnalysisStats(AnalysisStats),
    RunTests(RunTests),
    RustcTests(RustcTests),
    Diagnostics(Diagnostics),
    UnresolvedReferences(UnresolvedReferences),
    Ssr(Ssr),
    Search(Search),
    Lsif(Lsif),
    Scip(Scip),
}
```

`eldoc-box` displays the complete documentation, including size hint and alignment. I am expecting to see all the info displayed by `eldoc-box` in the echo area by eglot.

![2024-11-26_20-24-53_screenshot](https://github.com/user-attachments/assets/cc4a7686-5f23-49ff-a863-f0a3973b1c2e)

I first checked if rust-analyzer responds with incomplete info. Turns out rust-analyzer is innocent. 

    [jsonrpc] e[20:41:28.501] <-- textDocument/hover[27]
    (:jsonrpc "2.0" :id 27 :result
              (:contents
               (:kind "markdown" :value "\n```rust\nrust_analyzer::cli::flags::RustAnalyzerCmd\n```\n\n```rust\n// size = 56 (0x38), align = 0x8\nRustcTests(RustcTests)\n```")
               :range
               (:start
                (:line 92 :character 32)
                :end
                (:line 92 :character 42))))

So I figured it could be either `eldoc` or `eglot` that chooses to only show the first line.

After glimpsing through the eldoc package, especially functions like `eldoc-documentation-compose`, and variables like `eldoc-documentation-function`, plus some edebugging, it seems to me `eldoc-hover-eldoc-function` is called by `eldoc-documentation-compose`, where `eldoc-hover-eldoc-function` asynchronously calls the callback provided by `eldoc-documentation-compose`.

The interesting part is `eldoc-hover-eldoc-function` specifies the value to `:echo` as the first index of newline in `info`. Now the problem becomes what `info` is when the callback is invoked. Being a emacs-lisp beginner, I printed `info` to the `*Message*` buffer, and can verify `info` contains all the information rust-analyzer sends back. (I tried use edebug to see the value of `info`, but `eglot--lambda` is a macro, and edebugging a macro takes more work so I took the easy route. If you have better debugging technique, and would be so kind to inform me, please do :))

This is the modified version of `eldoc-hover-eldoc-function`
```lisp
(defun eglot-hover-eldoc-function (cb)
  "A member of `eldoc-documentation-functions', for hover."
  (when (eglot-server-capable :hoverProvider)
    (let ((buf (current-buffer)))
      (jsonrpc-async-request
       (eglot--current-server-or-lose)
       :textDocument/hover (eglot--TextDocumentPositionParams)
       :success-fn (eglot--lambda ((Hover) contents range)
                     (eglot--when-buffer-window buf
                       (let ((info (unless (seq-empty-p contents)
                                     (eglot--hover-info contents range))))
                         (message "Check info: %s" info) ; <--- Added logging.
                         (funcall cb info
                                  :echo (and info (string-match "\n" info))))))
       :deferred :textDocument/hover))
    (eglot--highlight-piggyback cb)
    t))
```

![2024-11-26_22-50-26_screenshot](https://github.com/user-attachments/assets/705421e0-72dc-4b5a-b160-1b12631f653e)

After replacing `:echo (and info (string-match "\n" info))` to `:echo info`, everything work as expected!

Now I am curious why the choice is made to only display the doc string until the first newline?

Finally, many thanks for your great package! I use it everyday, and I mean to study it to get more familiar with the LSP protocol, and maybe contribute to eglot someday :).

**Comment by joaotavora** [comment]
The echo area is a hotly contested small space.  Many eldoc documentation functions live there, and eldoc will ask for the "oneliner" `:echo` property of each if it is present.  So this is by design.  If you're cool with a particular documentation function to be a fat space hog, then look at this discussion for a customization https://github.com/joaotavora/eglot/discussions/1458

**Comment by joaotavora** [comment, marked as answer]
You should use something like this (taken from that discussion):

```lisp
(defun my/special-eglot-hover-function (cb)
  "Same as `eglot-hover-eldoc-function`, but throw away its short :echo cookie"
  (eglot-hover-eldoc-function (lambda (info &rest _ignore)
                                ;; ignore the :echo cookie that eglot-hover-eldoc-function offers
                                (funcall cb info))))

(add-hook
 'eglot-managed-mode-hook
 (lambda ()
   (setq-local eldoc-documentation-functions 
               (cl-substitute #'my/special-eglot-hover-function
                              'eglot-hover-eldoc-function 
                              eldoc-documentation-functions)))

**Comment by joaotavora** [comment]
Of course, remove your customization of `eldoc-documentation-functions` for the above to work, or at least don't set it to a non-list (that's deprecated).  Notice the plural "_functions_"

**Comment by zino299792** [comment]
Thanks for your quick reply. I tried your suggesion and it works and is much cleaner than redefining the function.

**Comment by richcarl** [comment]
The design decision might be for the best, but it would be good if the documentation (ch 4, Customizing Eglot) was updated. Currently it just says "to configure the amount of space taken up by documentation in the echo area, customize the ElDoc variable eldoc-echo-area-use-multiline-p". I lost hours trying to find why this doesn't actually work, and the only explanation seems to be in this discussion.

**Reply by joaotavora** [reply]
That sentence occurs with some sentence before it.  I have emphasized it. I can understand that you have not correctly interpreted it:

> To have Eglot behave like you want it, **sometimes it suffices to customize Emacs packages and features that Eglot uses but to do so for all of Emacs (regardless of Eglot) .** For example, to configure the amount of space taken up by documentation in the echo area, customize the ElDoc variable eldoc-echo-area-use-multiline-p. 

So, when you customize an ElDoc variable you are just customizing ElDoc *not* Eglot. And _"sometimes [that] suffices"_.  And sometimes, it doesn't, like here.  Here's some context.

Usually  the complaints I had received had been about *too much* space being used by ElDoc, and that's why that sentence is there with prominence.  They want ElDoc to absolutely *not* grow above one line (because they don't like what they call "jumping"). 

In this issue, though, it seems that 100% of the people are having the opposite problem.  And that, my friends, is called selection bias. :-)

In the most recent version of the manual, which you may find [here](https://elpa.gnu.org/devel/doc/eglot.html#More-Customization) there is a "More Customization" subsection, that briefly discusses with Eglot's ElDoc functions.   However, the specific request for this issue is not covered there, the more common  "ElDoc takes up too much space is" case is.  Patches are welcome to Emacs's master's `doc/misc/eglot.texi` to:

* put the initial sentence in (even better) context
* reorganize that subsection so that the concerns of the users of this issue are equally well represented 

It's not easy to write a manual, but it's not that hard either and by helping me do it you are helping others.  Submit patches to bug-gnu-emacs@gnu.org.

In the meantime I will try to link the most recent version of the manual in the homepage of this repo.
